Integrate CDC into k8ssandra-operator reconciliation logic and APIs

Miles-Garnsey opened the pull request:
**What this PR does**:

What it says on the tin.

Note that I've modified go.mod to use my recent changes to cass-operator. We need to revert this back to using a non-snapshot release prior to merge, but this is required because the cass-operator changes aren't yet merged.

**Which issue(s) this PR fixes**:
Fixes #570 

**Checklist**
- [ ] Changes manually tested
- [ ] Automated Tests added/updated
- [ ] Documentation added/updated
- [ ] CHANGELOG.md updated (not required for documentation PRs)
- [ ] CLA Signed:  [DataStax CLA](https://cla.datastax.com/)

adejanovski commented on apis/k8ssandra/v1alpha1/k8ssandracluster_types.go:
```diff
@@ -308,6 +308,11 @@ type DatacenterOptions struct {
 	Telemetry *telemetryapi.TelemetrySpec `json:"telemetry,omitempty"`
+
+	// CDC defines the desired state for CDC integrations. It can be used to feed mutation events from Cassandra into an Apache Pulsar cluster,
+	// from where they can be expored to external systems.
```
```suggestion
	// from where they can be exported to external systems.
```

adejanovski commented on pkg/cassandra/datacenter.go:
```diff
@@ -331,6 +333,10 @@ func Coalesce(clusterName string, clusterTemplate *api.CassandraClusterTemplate,
 		dcConfig.MgmtAPIHeap = dcTemplate.DatacenterOptions.MgmtAPIHeap
 	}
 
+	if dcTemplate.CDC != nil {
+		dcConfig.CDC = dcTemplate.CDC
+	}
```
```suggestion
	if dcTemplate.CDC == nil {
		dcConfig.CDC = clusterTemplate.CDC
	} else {
	        dcConfig.CDC = dcTemplate.CDC
	}
```

Miles-Garnsey replied:
No, no. I intended exactly what I'd written. We don't allow definition of CDC feeds cluster-wide, they are defined per DC. Think about where you want the data to flow (and how many copies you want to land in each Pulsar cluster) and you'll understand why.

adejanovski reviewed (changes requested):
The coalesce logic between cluster and dc templates doesn't seem to be applied as it should. Let me know if I'm missing something.

Otherwise, LGTM

adejanovski reviewed (approved):
All good then 👍

Miles-Garnsey commented:
**NOTE: DO NOT MERGE THIS PR** until changes are merged into cass-operator and we can reference a stable cass-operator version in `go.mod`.

Miles-Garnsey commented:
I have bumped the cass-operator version in go.mod to point at the current master. I have also rebased.

Pending the tests passing (it was a gnarly rebase) I think this is ready for merge. @adejanovski do you agree?

adejanovski commented:
> I have bumped the cass-operator version in go.mod to point at the current master. I have also rebased.
> 
> Pending the tests passing (it was a gnarly rebase) I think this is ready for merge. @adejanovski do you agree?

Could you instead rebase on top of main? The cass-operator 1.12.0 upgrade was just merged there.

Miles-Garnsey commented:
> Could you instead rebase on top of main? The cass-operator 1.12.0 upgrade was just merged there.

@adejanovski that's what I'm saying, this is rebased on main. I'm now seeing the cert-manager issues you've been reporting though since that rebase.

adejanovski commented:
> > Could you instead rebase on top of main? The cass-operator 1.12.0 upgrade was just merged there.
> 
> @adejanovski that's what I'm saying, this is rebased on main. I'm now seeing the cert-manager issues you've been reporting though since that rebase.

Nope, you're using the latest `main` [commit](https://github.com/k8ssandra/k8ssandra-operator/pull/580/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R15), when we want `v1.12.0` specifically until we cut out the release. While you can argue it's the same commit, it feels cleaner that the 1.2 release references cass-operator v1.12.0 specifically.

Miles-Garnsey commented:
> Nope, you're using the latest main [commit](https://github.com/k8ssandra/k8ssandra-operator/pull/580/files#diff-33ef32bf6c23acb95f5902d7097b7a1d5128ca061167ec0716715b0b9eeaa5f6R15), when we want v1.12.0 specifically until we cut out the release. While you can argue it's the same commit, it feels cleaner that the 1.2 release references cass-operator v1.12.0 specifically.

I agree that the release should use a real version instead of some commit on master. Since 1.12 is now released I've updated this PR again to point at it, and rebased again to try to resolver these cert-manager issues we've suddenly started seeing.

If the tests pass now this PR should be good to go.

adejanovski commented:
@Miles-Garnsey, we're still good for a merge and CI is all green (yay).
I'll merge this later today if you don't get a chance to.
